pushsync/pusher.go

```diff
@@ -0,0 +1,353 @@
+type pushedItem struct {
+	tag         *chunk.Tag       // tag for the chunk
+	shortcut    bool             // if the chunk receipt was sent by self
+	firstSentAt time.Time        // first sent at time
+	lastSentAt  time.Time        // most recently sent at time
+	synced      bool             // set when chunk got synced
+	span        opentracing.Span // roundtrip span
+}
+
+// NewPusher contructs a Pusher and starts up the push sync protocol
+// takes
+// - a DB interface to subscribe to push sync index to allow iterating over recently stored chunks
+// - a pubsub interface to send chunks and receive statements of custody
+// - tags that hold the tags
+func NewPusher(store DB, ps PubSub, tags *chunk.Tags) *Pusher {
+	p := &Pusher{
+		store:         store,
+		tags:          tags,
+		quit:          make(chan struct{}),
+		closed:        make(chan struct{}),
+		pushed:        make(map[string]*pushedItem),
+		receipts:      make(chan []byte),
+		ps:            ps,
+		logger:        log.New("self", hex.EncodeToString(ps.BaseAddr())),
+		retryInterval: 100 * time.Millisecond,
+	}
+	go p.sync()
+	return p
+}
+
+// Close closes the pusher
+func (p *Pusher) Close() {
+	close(p.quit)
+	<-p.closed
+}
+
+// sync starts a forever loop that pushes chunks to their neighbourhood
+// and receives receipts (statements of custody) for them.
+// chunks that are not acknowledged with a receipt are retried
+// not earlier than retryInterval after they were last pushed
+// the routine also updates counts of states on a tag in order
+// to monitor the proportion of saved, sent and synced chunks of
+// a file or collection
+func (p *Pusher) sync() {
+	var chunks <-chan chunk.Chunk
+	var unsubscribe func()
+	var syncedAddrs []storage.Address
+	var syncedItems []*pushedItem
+	defer close(p.closed)
+	// timer, initially set to 0 to fall through select case on timer.C for initialisation
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	// register handler for pssReceiptTopic on pss pubsub
+	deregister := p.ps.Register(pssReceiptTopic, false, func(msg []byte, _ *p2p.Peer) error {
+		return p.handleReceiptMsg(msg)
+	})
+	defer deregister()
+
+	chunksInBatch := -1
+	var batchStartTime time.Time
+	ctx := context.Background()
+
+	var average uint64 = 100000 // microseconds
+	var measurements uint64
+
+	for {
+		select {
+
+		// retry interval timer triggers starting from new
+		case <-timer.C:
+			// initially timer is set to go off as well as every time we hit the end of push index
+			// so no wait for retryInterval needed to set  items synced
+			metrics.GetOrRegisterCounter("pusher.subscribe-push", nil).Inc(1)
+			// if subscribe was running, stop it
+			if unsubscribe != nil {
+				unsubscribe()
+			}
+
+			syncedTags := make(map[uint32]int)
+			// set chunk status to synced, insert to db GC index
+			if err := p.store.Set(ctx, chunk.ModeSetSync, syncedAddrs...); err != nil {
+				panic(fmt.Sprintf("pushsync: error setting chunks to synced: %v", err))
```

Comment:
Why do we panic the whole process if there is a problem with this `Set`. I don't think we do that anywhere else. Maybe just log the error and exit from the forever loop?

Reply:
you are right but maybe we should panic elsewhere too if db errors no?
i dont think exiting the loop leads to any good

Reply:
We spoke about enabling push sync in the first release after it is merged as an experimental feature - one that you enable with a flag.

If push sync fails, because of some issue with it or PSS, the Swarm node can still continue to work with the regular pull sync for example, or at least continue to fetch content.

Reply:
but this is about a database failure. somewhat orthogonal to what you are saying.